exonum/src/runtime/rust/mod.rs

```diff
@@ -0,0 +1,174 @@
+
+#[macro_use]
+pub mod service;
+
+#[cfg(test)]
+pub mod tests;
+
+use super::{
+    error::{DeployError, ExecutionError, InitError},
+    ArtifactSpec, CallInfo, DeployStatus, EnvContext, InstanceInitData, RuntimeEnvironment,
+    ServiceInstanceId,
+};
+
+use crate::crypto::{Hash, PublicKey};
+use crate::storage::Fork;
+
+use self::service::SystemService;
+
+#[derive(Debug, Default)]
+struct RustRuntime {
+    inner: RefCell<RustRuntimeInner>,
+}
+
+impl RustRuntime {
+    fn add_service(&self, artifact: RustArtifactSpec, service: Box<dyn SystemService>) {
+        self.inner.borrow_mut().services.insert(artifact, service);
+    }
+}
+
+#[derive(Debug, Default)]
+struct RustRuntimeInner {
+    services: HashMap<RustArtifactSpec, Box<dyn SystemService>>,
+    deployed: HashSet<RustArtifactSpec>,
+    initialized: HashMap<ServiceInstanceId, Box<dyn SystemService>>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct RustArtifactSpec {
+    pub name: String,
+    pub version: (u32, u32, u32),
+}
+
+impl RuntimeEnvironment for RustRuntime {
+    fn start_deploy(&self, artifact: ArtifactSpec) -> Result<(), DeployError> {
+        let artifact = if let ArtifactSpec::Rust(artifact) = artifact {
+            artifact
+        } else {
+            return Err(DeployError::WrongArtifact);
+        };
+
+        let mut inner = self.inner.borrow_mut();
+
+        if !inner.services.contains_key(&artifact) {
+            return Err(DeployError::FailedToDeploy);
+        }
+
+        if !inner.deployed.insert(artifact) {
+            return Err(DeployError::AlreadyDeployed);
+        }
+        Ok(())
+    }
+
+    fn check_deploy_status(&self, artifact: ArtifactSpec) -> Result<DeployStatus, DeployError> {
+        let artifact = if let ArtifactSpec::Rust(artifact) = artifact {
+            artifact
+        } else {
+            return Err(DeployError::WrongArtifact);
+        };
+
+        let inner = self.inner.borrow();
+
+        if inner.deployed.contains(&artifact) {
+            Ok(DeployStatus::Deployed)
+        } else {
+            Err(DeployError::FailedToDeploy)
+        }
+    }
+
+    fn init_service(
+        &mut self,
+        _: &mut EnvContext,
+        artifact: ArtifactSpec,
+        init: &InstanceInitData,
+    ) -> Result<(), InitError> {
+        let artifact = if let ArtifactSpec::Rust(artifact) = artifact {
+            artifact
+        } else {
+            return Err(InitError::WrongArtifact);
+        };
+
+        let mut inner = self.inner.borrow_mut();
+
+        if !inner.deployed.contains(&artifact) {
+            return Err(InitError::NotDeployed);
+        }
+
+        if inner.initialized.contains_key(&init.instance_id) {
+            return Err(InitError::ServiceIdExists);
+        }
+
+        let service = inner.services.remove(&artifact).unwrap();
+        inner.initialized.insert(init.instance_id, service);
+        Ok(())
+    }
+
+    fn execute(
+        &self,
+        context: &mut EnvContext,
```

Comment:
EnvContext sounds a little bit unclear, imo `Environment` or `RuntimeContext` is clearly.